base/reduce.jl

```diff
@@ -1246,3 +1264,102 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
     end
     return n
 end
+
+# simd optimization for min/max related reduction
+# general fallback
+_fast(op, x, y) = op(x, y)
+_fast(op) = (x, y) -> _fast(op, x, y)
+
+# The following are used in optimized `mapreduce_impl` for IEEEFloat.
+_fast(::typeof(min), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(x-y), x, y)
+_fast(::typeof(max), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(y-x), x, y)
+
+# Some help function for better vectorization
+# NaN check
+_anynan(x, y) = isnan(x) | isnan(y)
+_anynan(x::NTuple{N}) where {N} = reduce(|, map(_anynan, x[1:N>>1], x[1+N>>1:N]))
+# split then reduce for inputs whose length > 4
+_half_reduce(op, x::NTuple{N}) where {N} = _half_reduce(op, map(op, x[1:N>>1], x[1+N>>1:N]))
+_half_reduce(op, x::NTuple{4}) = reduce(op, x)
+
+function mapreduce_impl(f, op::Union{typeof(max),typeof(min)},
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
+    @inline elf(i::Int) = @inbounds f(A[i])
+    Eltype = _return_type(elf, Tuple{Int})
+    # Limit this optimization to IEEEFloat. (general fallback is 2x faster for Integer)
+    Eltype <: IEEEFloat ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
+    @noinline firstnan(temp::Tuple) = temp[findfirst(isnan, temp)], last
+    function simd_kernal(::Val{N}, ini) where {N}
+        # reused kernal for better vectorizable reduction.
+        # - `N` is the unroll size.
+        # - `ini` is the initial value of reduction
+        # It returns the reduced result (`v`), and the last processed index (`i`).
+        isnan(ini) && return ini, last
+        vs = ntuple(Returns(ini), Val(N))
+        index = ntuple(identity, Val(N))
+        local i = first
+        while i <= last - N
+            temp = map(elf, i .+ index)
+            _anynan(temp) && return firstnan(temp) # nan check
+            vs = map(_fast(op), vs, temp) # then _fast(op) is safe
```

Comment:
It seems constructing a tuple like this can hurt performance when we have something like `Eltype === Union{Float32,Float64}`. We either need a manual promotion or restrict `Eltype` to a concrete one.